compiler/src/iree/compiler/Codegen/Common/GPU/GPUConvertToCoalescedDMA.cpp

```diff
@@ -35,243 +39,351 @@ namespace mlir::iree_compiler {
-                              /*useWarpMapping=*/false);
-    auto innerForallOp = scf::ForallOp::create(
-        rewriter, loc, innerLowerBounds, innerUpperBounds, threadTileSizes,
-        destWgSlice.getResult(), rewriter.getArrayAttr(innerMapping));
-
-    // Build the body of inner forall with DMA op
-    Block *innerBody = innerForallOp.getBody();
-    rewriter.setInsertionPointToStart(innerBody);
-
-    // Compute the lane ID by linearizing the thread IVs
-    SmallVector<Value> threadIVs(innerBody->getArguments().begin(),
-                                 innerBody->getArguments().begin() + rank);
-    Value laneId = affine::AffineLinearizeIndexOp::create(
-        rewriter, loc, threadIVs, *staticThreadTileSizes);
-
-    Value innerSharedOut = innerBody->getArguments()[rank];
-    auto innerInParallelOp =
-        cast<scf::InParallelOp>(innerBody->getTerminator());
-    Block &innerInParallelBlock = innerInParallelOp.getRegion().front();
-    rewriter.setInsertionPointToStart(&innerInParallelBlock);
-
-    IREE::GPU::CoalescedGatherDMAOp::create(
-        rewriter, loc, innerSharedOut.getType(), source,
-        ValueRange(vectorIndices), innerSharedOut, laneId);
-
-    // Insert the result of inner forall back into outer forall's output.
-    rewriter.setInsertionPointAfter(innerForallOp);
-    auto outerInParallelOp =
-        cast<scf::InParallelOp>(outerBody->getTerminator());
-    Block &outerInParallelBlock = outerInParallelOp.getRegion().front();
-    rewriter.setInsertionPointToStart(&outerInParallelBlock);
-
-    tensor::ParallelInsertSliceOp::create(
-        rewriter, loc, innerForallOp.getResult(0), outerSharedOut, outerOffsets,
-        subgroupTileSizes, strides);
-
-    rewriter.replaceOp(gatherOp, outerForallOp.getResults());
-    return success();
+    scf::ForallOp threadForallOp =
+        tileToThreadLevel(op, rewriter, threadNumThreads);
+    if (!threadForallOp) {
+      return failure();
+    }
+
+    return createDMAInForall<OpTy>(threadForallOp, rewriter);
+  }
+
+protected:
+  /// Compute thread num threads for the operation.
+  virtual SmallVector<OpFoldResult> computeThreadNumThreads(OpBuilder &builder,
+                                                            OpTy op) const = 0;
+};
+
+struct ConvertCopyToCoalescedDMA
+    : public ConvertToCoalescedDMABase<linalg::CopyOp> {
+  using ConvertToCoalescedDMABase<linalg::CopyOp>::ConvertToCoalescedDMABase;
+
+protected:
+  SmallVector<OpFoldResult>
+  computeThreadNumThreads(OpBuilder &builder,
+                          linalg::CopyOp copyOp) const override {
+    auto outputType = cast<RankedTensorType>(copyOp.getOutputs()[0].getType());
+    return computeThreadNumThreadsImpl(builder, copyOp, outputType);
+  }
+};
+
+struct ConvertGatherToCoalescedDMA
+    : public ConvertToCoalescedDMABase<IREE::LinalgExt::GatherOp> {
+  using ConvertToCoalescedDMABase<
+      IREE::LinalgExt::GatherOp>::ConvertToCoalescedDMABase;
```

Comment:
One note: I remember hearing from someone that `linalg_ext.gather` is actually an op that does the `tensor<[result dims]x[rank of source]xindex>` thing, so you can create your index slices by `extract_slice`ing the indices tensor. You'll probably want to double-check, and CC @IanWood1 who I think told me this